concourse/client/util.py

```diff
@@ -122,3 +152,159 @@ def jobs_not_triggered(
                     yield (job, pr_resource, pr_resource_version)
+
+
+def get_resource_lock_and_trigger_build(
+    job: Job,
+    resource: PipelineConfigResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+    max_retries: int=3,
+):
+    retries = max_retries
+    while retries >= 0:
+        try:
+            with pin_resource(
+                resource=resource,
+                resource_version=resource_version,
+                concourse_api=concourse_api,
+            ):
+                concourse_api.trigger_build(
+                    pipeline_name=resource.pipeline_name(),
+                    job_name=job.pipeline_name,
+                )
+
+                job_retry = 10
+                job_sleep_seconds = 5
+                if wait_for_job_to_be_triggered(
+                    job=job,
+                    resource_version=resource_version,
+                    concourse_api=concourse_api,
+                    retries=job_retry,
+                    sleep_time_seconds=job_sleep_seconds,
+                ):
+                    info(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        'has been triggered'
+                    )
+                else:
+                    pin_comment = PinComment(
+                        pin_timestamp=datetime.now().isoformat(),
+                        version_data=resource_version.version(),
+                        next_retry=(datetime.now() + timedelta(
+                            seconds=(job_retry + 1) * job_sleep_seconds,
+                        )).isoformat(),
+                        comment="pinned by technical user, please do not unpin manually"
+                    )
+                    concourse_api.pin_comment(
+                        pipeline_name=resource.pipeline_name(),
+                        resource_name=resource.name,
+                        comment=json.dumps(dataclasses.asdict(pin_comment)),
+                    )
+                    warning(
+                        f'job {job.name} for resource version {resource_version.version()} '
+                        f'could not be triggered. Will retry {retries} more time(s)'
+                    )
+                return
+        except PinningFailedError as e:
+            warning(e)
+            retries -= 1
+        except PinningUnnecessaryError as e:
+            # already beeing processed by another thread, nothing to do
+            info(e)
+            return
+
+    warning(
+        f'job {job.name} for resource version {resource_version.version()} '
+        f'could not be triggered. Giving up after {max_retries} retries'
+    )
+
+
+@contextmanager
+def pin_resource(
+    resource: PipelineResource,
+    resource_version: ResourceVersion,
+    concourse_api,
+):
+    _ensure_resource_unpinned(
+        resource=resource,
+        resource_version=resource_version,
+        concourse_api=concourse_api,
+        retries=10,
```

Comment:
I think the sleep-time + retries-amount should no longer be required to be passed; instead, use the existing metadata, as elaborated in a different comment yesterday